src/kbmod/standardizers/standardizer.py

```diff
@@ -0,0 +1,589 @@
+                 boxes or a list of `None`s if they can not be calculated.
+        ======== ==============================================================
+
+        Returns
+        -------
+        metadata : `dict`
+            Standardized metadata.
+
+        Notes
+        -----
+        If a target being standardized contains only 1 image that can be
+        standardized and processed by KBMOD, the values in the dictionary can
+        be strings, integers, floats, and other such single-values non-iterable
+        types, or an iterable type containing only 1 element.
+        When the target being standardized contains multiple images/extensions
+        that are processable by KBMOD the single-valued dict entries will be
+        considered as shared by all the processable images/extension/units of
+        the target. The iterable values in the dictionary must match the number
+        and order of the units marked as processable.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def standardizeScienceImage(self):
+        """Standardizes the science image data.
+
+        For FITS files, this is usually a trivial no-op operation that returns
+        the `data` attribute of the correct HDU. For different data sources
+        this is a type casting, or a download operation.
+
+        Returns
+        -------
+        image : `list[~np.array]`
+            Science images.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def standardizeVarianceImage(self):
+        """Standardizes the variance data.
+
+        For FITS files, this is sometimes a trivial no-op operation returning
+        the correct FITS extension. In other cases, this has to be calculated
+        if sufficient information is provided in the header or a different
+        file needs to be downloaded/read.
+
+        .. note::
+           Refer to the manual for the originating dataset, whether the
+           instrument or processing pipeline reference, as some variance planes
+           store the inverse of variance.
+
+        Returns
+        -------
+        variance : `list[`np.array]`
+            Variance images.
+        """
+        raise NotImplementedError()
+
+    @abc.abstractmethod
+    def standardizeMaskImage(self):
+        """Standardizes the mask data as an simple 0 (not masked) and 1
+        (masked) bitmap.
+
+        For FITS files, this is sometimes a trivial no-op operation returning
+        the correct FITS extension. In other cases, the mask has to be
+        constructed from external data, such as pixel masks and catalogs,
+        or downloaded/read from a different file at the data source.
+
+        Returns
+        -------
+        mask : `list[~np.array]`
+            Mask images.
+        """
+        raise NotImplementedError()
+
+    # no idea really what to do bout this one? AFAIK almost no images come
+    # with PSFs in them
```

Comment:
What is Rubin's plan for providing PSFs? Is is a separate query?

Reply:
It's in the FITS file (that is, it's part of the object you can retrieve via the Butler), but it's their own internal representation of a PSF - so this isn't something we can just rely on as a generic thing. We can make the `ButlerStandardizer` use it to load up the object in Python - but then we need to do something with that object, like evaluate a realization of a PSF at some wavelength and position on the CCD as a numpy array for example. But even then we need to line up that realization with pixel size basically, so that the realization is basically over the same pixels as our images, i.e. so their physical sizes are comparable, because that's how the current PSF class works - the kernel dimension is directly related to the pixel size.  For other instruments I've no idea what to do tbh.

A part of the problem is that PSF class we have is not a good representation of a PSF. Not because Gaussians are bad functional representation of a PSFm but because it only stores a representation of a PSF as an array and not some functional expression.
But even then, as far as I have seen there are no ways the [Rubin PSF class](https://github.com/lsst/afw/blob/main/include/lsst/afw/detection/Psf.h#L75) can be "exported" as any of the to extract the functional PSF forms from AstroPy (f.e. like [moffat](https://docs.astropy.org/en/stable/api/astropy.convolution.Moffat2DKernel.html)) or [photutils](https://photutils.readthedocs.io/en/stable/psf.html) - which again wouldn't even help us that much because we still need an C++ representation of the same again if we want to evaluate it at points (in a way that is generic to where we can do things like PSFs that vary across the CCD) and so on. I guess we can evaluate it and then fit it - but the costs involved here are kinda crazy.